File: Backtracking/90.subsets-ii.py
from typing import List


# @leet start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n, path, res = len(nums), [], []
        nums.sort()

        def dfs(start):
            res.append(path[:])
            for i in range(start, n):
                if i > start and nums[i] == nums[i - 1]:
                    continue
                path.append(nums[i])
                dfs(i + 1)
                path.pop()

        dfs(0)
        return res


# @leet end

File: Backtracking/90.subsets-ii.ts
// @leet start
function subsetsWithDup(nums: number[]): number[][] {
  const res: number[][] = [];
  const path: number[] = [];
  function dfs(start: number) {
    res.push(path.slice());
    for (let i = start; i < nums.length; i++) {
      if (i > start && nums[i] === nums[i - 1]) {
        continue;
      }
      const v = nums[i];
      path.push(v);
      dfs(i + 1);
      path.pop();
    }
  }
  dfs(0);
  return res;
}
// @leet end
